JetNews/app/src/main/java/com/example/jetnews/ui/home/HomeViewModel.kt

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.jetnews.ui.home
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.viewModelScope
+import com.example.jetnews.R
+import com.example.jetnews.data.Result
+import com.example.jetnews.data.posts.PostsRepository
+import com.example.jetnews.model.Post
+import com.example.jetnews.utils.ErrorMessage
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import java.util.UUID
+
+/**
+ * UI state for the Home screen
+ */
+data class HomeUiState(
+    val posts: List<Post> = emptyList(),
+    val favorites: Set<String> = emptySet(),
+    val loading: Boolean = false,
+    val errorMessages: List<ErrorMessage> = emptyList()
+) {
+    /**
+     * True if this represents a first load
+     */
+    val initialLoad: Boolean
+        get() = posts.isEmpty() && favorites.isEmpty() && errorMessages.isEmpty() && loading
+}
+
+/**
+ * ViewModel that handles the business logic of the Home screen
+ */
+class HomeViewModel(
+    private val postsRepository: PostsRepository
+) : ViewModel() {
+
+    // UI state exposed to the UI
+    private val _uiState = MutableStateFlow(HomeUiState(loading = true))
+    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()
+
+    init {
+        refreshPosts()
+
+        // Observe for favorite changes in the repo layer
+        viewModelScope.launch {
+            postsRepository.observeFavorites().collect { favorites ->
+                _uiState.update { it.copy(favorites = favorites) }
+            }
+        }
+    }
+
+    /**
+     * Refresh posts and update the UI state accordingly
+     */
+    fun refreshPosts() {
```

Comment:
How do you feel about deduplicating concurrent refreshes here?

Reply:
Unless it's 1-2 lines of simple code I wouldn't complicate things.